### PR TITLE
Initial MinIO CSV Loader

### PR DESCRIPTION
**Pull request by we-z**
### Summary of Changes

_Include a summary of changes then remove this line_

### Tests

_What tests did you add or modify and why? If no tests were added or modified, explain why. Remove this line_

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList

Make sure you have checked **all** steps below to ensure a timely review.

- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, **OR** documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes `make test`

### CODE REVIEW

File: cronjob.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  creationTimestamp: "2020-09-30T09:03:24Z"
+  name: minio-amundsen-loader
+  namespace: seldon
+  resourceVersion: "7409973"
+  selfLink: /apis/batch/v1beta1/namespaces/seldon/cronjobs/minio-amundsen-loader
+  uid: d1020f7e-67bc-4165-bff8-7bf357b1efa0
```
Comment (munkyshi) [inline cronjob.yml]: This shouldn't be relevant for a `kubectl apply -f`

File: cronjob.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  creationTimestamp: "2020-09-30T09:03:24Z"
```
Comment (munkyshi) [inline cronjob.yml]: This shouldn't be relevant for a `kubectl apply -f`

File: cronjob.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  name: minio-amundsen-loader
+  namespace: seldon
```
Comment (munkyshi) [inline cronjob.yml]: Why is the namespace `seldon`?

File: cronjob.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    metadata:
+      creationTimestamp: null
+      name: minio-amundsen-loader
+    spec:
+      template:
+        metadata:
+          creationTimestamp: null
+        spec:
+          containers:
+          - image: simplejob
```
Comment (munkyshi) [inline cronjob.yml]: Let's give the image a more descriptive name than `simplejob`

File: databuilder/extractor/minio_extractor.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# Copyright Contributors to the Amundsen project.
+# SPDX-License-Identifier: Apache-2.0
+
+import csv
+import importlib
+# from collections import defaultdict
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: nit: remove commented code

File: databuilder/extractor/minio_extractor.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+# Copyright Contributors to the Amundsen project.
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: nit: Let's avoid licensing our private code.

File: databuilder/extractor/minio_extractor.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+from pyhocon import ConfigTree
+from typing import Any
+import boto3
+
+from databuilder.extractor.base_extractor import Extractor
+from databuilder.models.table_metadata import TableMetadata, ColumnMetadata
+
+class MinioExtractor(Extractor):
+
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: nit: extra whitespace

File: databuilder/extractor/minio_extractor.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    An Extractor that extracts records via CSV.
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: Class comment should describe more than just CSV -- `via CSV` is also very vague.

File: databuilder/extractor/minio_extractor.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    def init(self, conf: ConfigTree) -> None:
+        """
+        :param conf:
+        """
+        self.conf = conf
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py:21]: What are we actually loading from the conf? Presumably the bucket name, endpoint url, aws_access_key_id, and aws_secret_access_key should all be configuration points that are passed from the original argparse.

File: databuilder/extractor/minio_extractor.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        """Get a list of keys in an S3 bucket."""
+        keys = []
+        resp = self.client.list_objects_v2(Bucket=bucket)
+        for obj in resp['Contents']:
+            if obj['Key'].split('/')[-1] == 'data.csv':
+                keys.append(obj['Key'])
+        return iter(keys)
+
+    def extract(self) -> Any:
+
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: nit: whitespace at top of file

File: databuilder/extractor/minio_extractor.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        self._extract_iter = None
+
+        self.client = boto3.client('s3',
+                          endpoint_url='http://dev-master:9000/',
+                          aws_access_key_id='myaccesskey',
+                          aws_secret_access_key='mysecretkey',
+                          region_name='us-east-1')
+
+    def get_s3_keys(self, bucket):
+        """Get a list of keys in an S3 bucket."""
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: Function name and comment are no longer accurate. Perhaps `get_data_csv_keys`?

File: databuilder/extractor/minio_extractor.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        if not self._extract_iter:
+            self._extract_iter = self.get_s3_keys('dev-raw-data')
+        try:
+            name = next(self._extract_iter)
+
+            r = self.client.select_object_content(
+                Bucket='dev-raw-data',
+                Key=name,
+                ExpressionType='SQL',
+                Expression="select * from s3object",
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: This seems bad performance-wise -- you're loading the entire csv to parse only the headers. Is there a cleaner way to do this?

File: databuilder/extractor/minio_extractor.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            table = TableMetadata(database='minio',
+                                  cluster='dev-raw-data',
+                                  schema='minio',
+                                  name=name.split('/',1)[0],
+                                  description='',
+                                  columns=colMetadatalist,
+                                  # TODO: this possibly should parse stringified booleans;
+                                  # right now it only will be false for empty strings
+                                  is_view=True,
+                                  tags='real-estate'
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: `real-estate` probably doesn't apply to most tables in minio. The two that I would suggest the minio extractor to automatically here are:
```
['minio', 'raw']
```

File: databuilder/extractor/minio_extractor.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+                                  )
+            return table
+        except StopIteration:
+            return None
+
+    def get_scope(self) -> str:
+        return 'extractor.minio.csv'
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: nit: newline at end of file (https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline)

File: example/scripts/minio_core_loader.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
+
+
```
Comment (munkyshi) [inline example/scripts/minio_core_loader.py]: Unnecessary whitespace.

Comment (munkyshi) [reply example/scripts/minio_core_loader.py]: Also, why is this file called `minio_core_loader`? The `yugabyte_core_loader` was loading Core files specifically, but in this case we're loading all sorts of CSVs in minio -- can we call this `minio_csv_loader`?

File: example/scripts/minio_core_loader.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+# todo: connection string needs to change
+def connection_string():
+    user = args.user
+    password = args.password
+    host = args.hostname
+    port = args.port
+    db = args.db
+    return "postgresql://%s:%s@%s:%s/%s" % (user, password, host, port, db)
+
+
```
Comment (munkyshi) [inline example/scripts/minio_core_loader.py]: This is dead code. Lets delete it.

File: example/scripts/minio_core_loader.py [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        'publisher.neo4j.neo4j_password': neo4j_password,
+        'publisher.neo4j.neo4j_encrypted': False,
+        'publisher.neo4j.job_publish_tag': 'unique_lastupdated_tag',  # should use unique tag here like {ds}
+    })
+
+    return DefaultJob(conf=job_config,
+                      task=task,
+                      publisher=Neo4jCsvPublisher())
+
+
```
Comment (munkyshi) [inline example/scripts/minio_core_loader.py]: Is this the exact same as in YugabyteSQLLoader? We should extract it to a base class to avoid duplicated logic.

File: example/scripts/minio_core_loader.py [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+    port = args.port
+    db = args.db
+    return "postgresql://%s:%s@%s:%s/%s" % (user, password, host, port, db)
+
+
+def run_minio_job():
+    where_clause_suffix = textwrap.dedent("""
+        where table_schema = 'public'
+    """)
+
```
Comment (munkyshi) [inline example/scripts/minio_core_loader.py]: This seems to be dead code.

**Review by munkyshi — CHANGES_REQUESTED** [review]

File: cronjob.yml [hunk 19]
```diff
@@ -1,12 +1,7 @@
 apiVersion: batch/v1beta1
 kind: CronJob
 metadata:
-  creationTimestamp: "2020-09-30T09:03:24Z"
   name: minio-amundsen-loader
-  namespace: seldon
```
Comment (munkyshi) [inline cronjob.yml]: IMO we should still be opinionated about a namespace -- would recommend `amundsen` as the namespace for this loader.

File: example/scripts/minio_csv_loader.py [hunk 20]
```diff
@@ -67,6 +67,14 @@ def run_minio_job():
...
 
     job_config = ConfigFactory.from_dict({
+        'extractor.minio.csv.{}'.format(MinioExtractor.ACCESS_KEY):
+            'myaccesskey',
+        'extractor.minio.csv.{}'.format(MinioExtractor.SECRET_KEY):
+            'mysecretkey',
+        'extractor.minio.csv.{}'.format(MinioExtractor.BUCKET_NAME):
+            'dev-raw-data',
+        'extractor.minio.csv.{}'.format(MinioExtractor.ENDPOINT_URL):
+            'http://dev-master:9000/',
```
Comment (munkyshi) [inline example/scripts/minio_csv_loader.py]: We shouldn't assume that anyone else has the alias dev-master -- also this should be based on `args.hostname`

File: databuilder/extractor/minio_extractor.py [hunk 21]
```diff
@@ -52,7 +52,7 @@ def extract(self) -> Any:
                 Bucket=self.bucket_name,
                 Key=name,
                 ExpressionType='SQL',
-                Expression="select * from s3object",
+                Expression="select * from s3object limit 1",
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py]: Can this be limit 0? I would expect it to still work.

File: databuilder/extractor/minio_extractor.py (hunk 7 above)
Comment (munkyshi) [reply databuilder/extractor/minio_extractor.py]: @we-z the extra whitespace still seems to be here?

File: cronjob.yml [hunk 22]
```diff
@@ -0,0 +1,32 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: minio-amundsen-loader
```
Comment (munkyshi) [inline cronjob.yml:4]: We should still be opinionated about namespace -- can we add `namespace: amundsen` here?

File: databuilder/extractor/minio_extractor.py [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+            table = TableMetadata(database='minio',
+                                  cluster='dev-raw-data',
+                                  schema='minio',
+                                  name=name.split('/',1)[0],
+                                  description='',
+                                  columns=colMetadatalist,
+                                  # TODO: this possibly should parse stringified booleans;
+                                  # right now it only will be false for empty strings
+                                  is_view=True,
```
Comment (munkyshi) [inline databuilder/extractor/minio_extractor.py:84]: Is this a view? TBH I'm not clear on what flag means.

File: Dockerfile [hunk 24]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.7-slim as base
+WORKDIR /app
+
+COPY requirements.txt /app/requirements.txt
+RUN pip3 install -r requirements.txt
+
+COPY . /app
+RUN python3 setup.py install
+
+ENTRYPOINT [ "python3",  "example/scripts/minio_core_loader.py" ]
```
Comment (munkyshi) [inline Dockerfile]: The entrypoint path needs to be updated to minio_csv_loader.py

File: Dockerfile [hunk 25]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.7-slim as base
```
Comment (munkyshi) [inline Dockerfile]: Can we rename this to minioloader.Dockerfile? Since we also have `ybloader.Dockerfile`?

File: example/scripts/minio_csv_loader.py [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+    node_files_folder = '{tmp_folder}/nodes/'.format(tmp_folder=tmp_folder)
+    relationship_files_folder = '{tmp_folder}/relationships/'.format(tmp_folder=tmp_folder)
+
+    job_config = ConfigFactory.from_dict({
+        'extractor.minio.csv.{}'.format(MinioExtractor.ACCESS_KEY):
+            'myaccesskey',
+        'extractor.minio.csv.{}'.format(MinioExtractor.SECRET_KEY):
+            'mysecretkey',
+        'extractor.minio.csv.{}'.format(MinioExtractor.BUCKET_NAME):
+            'dev-raw-data',
```
Comment (munkyshi) [inline example/scripts/minio_csv_loader.py:82]: Ideally these should be parameterized from the arg parser.

File: databuilder/extractor/minio_extractor.py (hunk 23 above)
Comment (we-z) [reply databuilder/extractor/minio_extractor.py:84]: No idea

**Review by munkyshi — APPROVED** [review]
@we-z Approved -- lets go ahead and squash + merge this bad boy. I'll throw together a follow up PR for the yugabyte loader cleanup bit.
